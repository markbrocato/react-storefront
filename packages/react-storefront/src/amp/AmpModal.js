/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Component, Fragment } from 'react'
import { Helmet } from 'react-helmet'
import withStyles from '@material-ui/core/styles/withStyles'
import PropTypes from 'prop-types'

export const styles = theme => ({
  root: {
    zIndex: theme.zIndex.amp.modal + 2, // zIndex of AppBar's withAmp + 1
    backgroundColor: 'rgba(255, 255, 255, .8)',
  },
})

/**
 * An AMP-compatible modal based on `<amp-lightbox>`
 * 
 * All props not specifically documented here are spread to the `<amp-lightbox>` element.
 * More info about amp-lightbox on AMP docs: https://www.ampproject.org/docs/reference/components/amp-lightbox
 * 
 * Usage example:
 * 
 *  <button on="tap:modal">Open Modal</button>
 * 
 *  <AmpModal id="modal">
 *    <button on="tap:modal.close">Close Modal</button>
 *    <div>Modal content ...</div>
 *  </AmpModal>
 */
@withStyles(styles, { name: 'RSFAmpModal' })
export default class AmpModal extends Component {

  static propTypes = {
    /**
     * The id for the amp-lightbox element.
     * This is REQUIRED attribute.
     */
    id: PropTypes.string,
    
    /**
     * Defines the style of animation for opening the lightbox. 
     * By default, this will be set to `fade-in`. 
     * Valid values are `fade-in`, `fly-in-bottom` and `fly-in-top`.
     * 
     * Note: The `fly-in-*` animation presets modify the transform property of the amp-lightbox element. 
     * Do not rely on transforming the amp-lightbox element directly. 
     * If you need to apply a transform, set it on a nested element instead. 
     */
    animateIn: PropTypes.string,
  }

  static defaultProps = {
    animateIn: 'fade-in',
  }

  render() {
    let { 
      id, 
      animateIn,
      classes, 
      children,
      ...others 
    } = this.props

    if (!id) {
      return new Error('Prop `id` is required for `AmpModal` component')
    }
    
    return (
      <Fragment>
        <Helmet key="helmet">
          <script async custom-element="amp-lightbox" src="https://cdn.ampproject.org/v0/amp-lightbox-0.1.js"></script>
        </Helmet>
        <amp-lightbox 
          key="lightbox" 
          layout="nodisplay"
          id={id}
          animate-in={animateIn}
          class={classes.root}
          {...others}
          >
          { children }
        </amp-lightbox>
      </Fragment>
    )
  }

}