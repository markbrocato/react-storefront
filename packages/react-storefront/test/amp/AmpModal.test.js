/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React from 'react'
import AmpModal from '../../src/amp/AmpModal'
import { Provider } from 'mobx-react'
import TestProvider from '../TestProvider'
import { mount } from 'enzyme'

describe('AmpModal', () => {
  it('should render', () => {
    expect(
      mount(
        <TestProvider>
          <Provider amp={true}>
            <AmpModal id="my-modal">
              <div>inner</div>
            </AmpModal>
          </Provider>
        </TestProvider>
      )
    ).toMatchSnapshot()
  })

  it('should render with custom props', () => {
    expect(
      mount(
        <TestProvider>
          <Provider amp={true}>
            <AmpModal 
              id="my-modal" 
              animateIn="fly-in-bottom" 
              on={`amp-modal-custom-id:AMP.setState({ someStateId: { modalOpened: true }})`}>
              <div>inner</div>
            </AmpModal>
          </Provider>
        </TestProvider>
      )
    ).toMatchSnapshot()
  })
})
