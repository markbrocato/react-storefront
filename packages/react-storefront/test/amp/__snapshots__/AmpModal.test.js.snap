// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmpModal should render 1`] = `
<TestProvider>
  <Provider router={{...}} app={{...}} history={{...}}>
    <MuiThemeProvider theme={{...}}>
      <Provider amp={true}>
        <WithStyles(AmpModal) id="my-modal">
          <AmpModal id="my-modal" classes={{...}} animateIn="fade-in">
            <HelmetWrapper defer={true} encodeSpecialCharacters={true}>
              <SideEffect(NullComponent) defer={true} encodeSpecialCharacters={true} script={{...}}>
                <NullComponent defer={true} encodeSpecialCharacters={true} script={{...}} />
              </SideEffect(NullComponent)>
            </HelmetWrapper>
            <amp-lightbox layout="nodisplay" id="my-modal" animate-in="fade-in" class="RSFAmpModal-root-1">
              <div>
                inner
              </div>
            </amp-lightbox>
          </AmpModal>
        </WithStyles(AmpModal)>
      </Provider>
    </MuiThemeProvider>
  </Provider>
</TestProvider>
`;

exports[`AmpModal should render with custom props 1`] = `
<TestProvider>
  <Provider router={{...}} app={{...}} history={{...}}>
    <MuiThemeProvider theme={{...}}>
      <Provider amp={true}>
        <WithStyles(AmpModal) id="my-modal" animateIn="fly-in-bottom" on="amp-modal-custom-id:AMP.setState({ someStateId: { modalOpened: true }})">
          <AmpModal id="my-modal" animateIn="fly-in-bottom" on="amp-modal-custom-id:AMP.setState({ someStateId: { modalOpened: true }})" classes={{...}}>
            <HelmetWrapper defer={true} encodeSpecialCharacters={true}>
              <SideEffect(NullComponent) defer={true} encodeSpecialCharacters={true} script={{...}}>
                <NullComponent defer={true} encodeSpecialCharacters={true} script={{...}} />
              </SideEffect(NullComponent)>
            </HelmetWrapper>
            <amp-lightbox layout="nodisplay" id="my-modal" animate-in="fly-in-bottom" class="RSFAmpModal-root-2" on="amp-modal-custom-id:AMP.setState({ someStateId: { modalOpened: true }})">
              <div>
                inner
              </div>
            </amp-lightbox>
          </AmpModal>
        </WithStyles(AmpModal)>
      </Provider>
    </MuiThemeProvider>
  </Provider>
</TestProvider>
`;
